### PR TITLE
[IOTDB-5368] add port check for confignode and datanode

### PR DESCRIPTION
**Pull request by l2280212**
add port check for confignode and datanode to avoid launching node which has same ports
see more information in [IOTDB-5368](https://issues.apache.org/jira/browse/IOTDB-5368)

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeStartupCheck.java [hunk 1]
```diff
@@ -31,18 +32,39 @@
...
 
   private static final ConfigNodeConfig CONF = ConfigNodeDescriptor.getInstance().getConf();
 
+  public ConfigNodeStartupCheck(String nodeRole) {
+    super(nodeRole);
+  }
+
+  private void portCheck() throws StartupException {
+    Set<Integer> portSet = new HashSet<>();
+    int configNodePort = 2;
```
Comment (SpriCoder) [inline confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeStartupCheck.java:54]: Please do nothing use Hard Code here. A good solution is use static final params

File: node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java [hunk 2]
```diff
@@ -43,19 +45,17 @@ public class StartupChecks {
...
+  protected StartupChecks(String nodeRole) {
+    this.nodeRole = nodeRole;
   }
 
   private void checkJMXPort(String nodeRole) {
     Boolean jmxLocal = Boolean.valueOf(System.getProperty(IoTDBConstant.IOTDB_JMX_LOCAL));
     String jmxPort = System.getProperty(IoTDBConstant.IOTDB_JMX_PORT);
 
-    if (jmxLocal) {
+    if (Boolean.TRUE.equals(jmxLocal)) {
```
Comment (SpriCoder) [inline node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java:58]: Why use equals here?

**Review by SpriCoder — CHANGES_REQUESTED** [review]
Good job. Some minor problems please take a look

File: node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java (hunk 2 above)
Comment (l2280212) [reply node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java:58]: This is a strange suggestion from sonarlint,(I also don't know why use equals. :(
I change it back

File: confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeStartupCheck.java (hunk 1 above)
Comment (l2280212) [reply confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeStartupCheck.java:54]: fixed

File: node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java [hunk 3]
```diff
@@ -55,7 +55,7 @@ private void checkJMXPort(String nodeRole) {
     Boolean jmxLocal = Boolean.valueOf(System.getProperty(IoTDBConstant.IOTDB_JMX_LOCAL));
```
Comment (OneSizeFitsQuorum) [inline node-commons/src/main/java/org/apache/iotdb/commons/service/StartupChecks.java]: use `boolean jmxLocal = Boolean.parseBoolean(System.getProperty(IoTDBConstant.IOTDB_JMX_LOCAL));`

File: server/src/main/java/org/apache/iotdb/db/conf/DataNodeStartCheck.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.iotdb.commons.service.StartupChecks;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/** DataNodeStartCheck checks the parameters in iotdb-datanode.properties when start and restart */
+public class DataNodeStartCheck extends StartupChecks {
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/conf/DataNodeStartCheck.java:31]: For consistent with the abstract class:
```suggestion
/** DataNodeStartupCheck checks the parameters in iotdb-datanode.properties when start and restart */
public class DataNodeStartupCheck extends StartupChecks {
```

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java [hunk 5]
```diff
@@ -297,4 +301,47 @@ public void testIllegalNodeRestart()
...
+          (MppBaseConfig) EnvFactory.getEnv().getConfig().getConfigNodeConfig(),
+          (MppCommonConfig) EnvFactory.getEnv().getConfig().getConfigNodeCommonConfig(),
+          null);
+      portConflictConfigNodeWrapper.start();
+      for (int i = 0; i < START_RETRY_NUM; ++i) {
+        showClusterResp = client.showCluster();
+        Thread.sleep(1000);
+      }
+      int afterStartConfigNodes = showClusterResp.getConfigNodeListSize();
+      Assert.assertEquals(beforeStartConfigNodes, afterStartConfigNodes);
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java:329]: Either assertEquals of ConfigNodes within each loop or just sleep 10s here.

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java [hunk 6]
```diff
@@ -297,4 +301,47 @@ public void testIllegalNodeRestart()
...
+          (MppBaseConfig) EnvFactory.getEnv().getConfig().getDataNodeConfig(),
+          (MppCommonConfig) EnvFactory.getEnv().getConfig().getDataNodeCommonConfig(),
+          null);
+      portConflictDataNodeWrapper.start();
+      for (int i = 0; i < START_RETRY_NUM; ++i) {
+        showClusterResp = client.showCluster();
+        Thread.sleep(1000);
+      }
+      int afterStartDataNodes = showClusterResp.getDataNodeListSize();
+      Assert.assertEquals(beforeStartDataNodes, afterStartDataNodes);
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java:344]: Either assertEquals of DataNodes within each loop or just sleep 10s here.

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL~

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java (hunk 6 above)
Comment (l2280212) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/conf/DataNodeStartCheck.java (hunk 4 above)
Comment (l2280212) [reply server/src/main/java/org/apache/iotdb/db/conf/DataNodeStartCheck.java]: fixed :)

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java [hunk 7]
```diff
@@ -195,7 +199,7 @@ public void testIllegalNodeRestart()
 
       TConfigNodeRestartReq configNodeRestartReq =
           ConfigNodeTestUtils.generateTConfigNodeRestartReq(
-              ERROR_CLUSTER_NAME, 1, registeredConfigNodeWrapper);
+              ERROR_CLUSTER_NAME, -1, registeredConfigNodeWrapper);
```
[the remark this answers is not in the document]
Comment (l2280212) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java]: In one test, I changed it and found that it did not affect the final result, so I forgot to change it back :(
Now fixed

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java [hunk 8]
```diff
@@ -297,4 +301,50 @@ public void testIllegalNodeRestart()
...
+          (MppBaseConfig) EnvFactory.getEnv().getConfig().getConfigNodeConfig(),
+          (MppCommonConfig) EnvFactory.getEnv().getConfig().getConfigNodeCommonConfig(),
+          null);
+      portConflictConfigNodeWrapper.start();
+      int afterStartConfigNodes;
+      for (int i = 0; i < START_RETRY_NUM; ++i) {
+        showClusterResp = client.showCluster();
+        Thread.sleep(1000);
+        afterStartConfigNodes = showClusterResp.getConfigNodeListSize();
+        Assert.assertEquals(beforeStartConfigNodes, afterStartConfigNodes);
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java:330]: The sleep statement is usually placed at last
```suggestion
        showClusterResp = client.showCluster();
        afterStartConfigNodes = showClusterResp.getConfigNodeListSize();
        Assert.assertEquals(beforeStartConfigNodes, afterStartConfigNodes);
        Thread.sleep(1000);
```

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java [hunk 9]
```diff
@@ -297,4 +301,50 @@ public void testIllegalNodeRestart()
...
+          (MppBaseConfig) EnvFactory.getEnv().getConfig().getDataNodeConfig(),
+          (MppCommonConfig) EnvFactory.getEnv().getConfig().getDataNodeCommonConfig(),
+          null);
+      portConflictDataNodeWrapper.start();
+      int afterStartDataNodes;
+      for (int i = 0; i < START_RETRY_NUM; ++i) {
+        showClusterResp = client.showCluster();
+        Thread.sleep(1000);
+        afterStartDataNodes = showClusterResp.getDataNodeListSize();
+        Assert.assertEquals(beforeStartDataNodes, afterStartDataNodes);
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java:346]: The sleep statement is usually placed at last
```suggestion
        showClusterResp = client.showCluster();
        afterStartDataNodes = showClusterResp.getDataNodeListSize();
        Assert.assertEquals(beforeStartDataNodes, afterStartDataNodes);
        Thread.sleep(1000);
```

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL~

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java (hunk 9 above)
Comment (l2280212) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterNodeErrorStartUpIT.java:346]: fixed

**Review by SpriCoder — APPROVED** [review]
LGTM~

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**